lib/base/timer.cpp

```diff
@@ -311,11 +326,20 @@ void Timer::TimerThreadProc()
 		 * until the current call is completed. */
 		l_Timers.erase(timer);
 
+		auto keepAlive (timer->m_Self.lock());
+
+		if (!keepAlive) {
+			// The last std::shared_ptr is gone, let ~Timer() proceed
+			continue;
+		}
```

Comment:
No, the destructor already waits for this block, don't worry.

Reply:
Both this block which fetches the next timer from the index and the destructor which removes the instance from the index use the same mutex.

Reply:
So basically `~Timer()`/`Stop()` ensures that while anything holds `l_TimerMutex`, that if a `Timer*` is currently stored inside `l_Timers`, then it's still safe to access its members (at least `m_Next` and `m_Self`), even if `~Timer()` already started executing?

In contrast, if `TimerHolder` would store a `std::weak_ptr<Timer>` instead of a `Timer*`, then the `weak_ptr` is part of `l_Timers` and accessing it is safe (again, guaranteed by `l_TimerMutex`, but this time without requiring elaborate reasoning extending the scope of that mutex to members of other classes even during their destruction). Then, to perform any access to the timer happens after a `lock()` on the `weak_ptr`, where the safety of this access is now guaranteed by the `weak_ptr`.

Do I miss a downside of that other approach? On the plus side, it seems to be much easier to reason for the safety of that approach.

Reply:
> So basically `~Timer()`/`Stop()` ensures that while anything holds `l_TimerMutex`, that if a `Timer*` is currently stored inside `l_Timers`, then it's still safe to access its members (at least `m_Next` and `m_Self`), even if `~Timer()` already started executing?

* Before: if Timer::TimerThreadProc() locks l_TimerMutex first, it toggles the next timer's (in l_Timers) m_Running. Once Timer#Stop() gets l_TimerMutex, it notices the latter and waits. If -in contrast- Timer#Stop() locks l_TimerMutex first, the instance disappears from l_Timers before Timer::TimerThreadProc() even notices what's happening.
* After: the scopes/codepaths of the lock are basically the same. But now instead of just toggling m_Running we check m_Self first. If the timer is still in l_Timers, then -as before- Timer#Stop() hasn't finished, Timer#~Timer() hasn't finished and weak_ptr#~weak_ptr() doesn’t even run yet.

> In contrast, if `TimerHolder` would store a `std::weak_ptr<Timer>` instead of a `Timer*`, then the `weak_ptr` is part of `l_Timers` and accessing it is safe (again, guaranteed by `l_TimerMutex`, but this time without requiring elaborate reasoning extending the scope of that mutex to members of other classes even during their destruction). Then, to perform any access to the timer happens after a `lock()` on the `weak_ptr`, where the safety of this access is now guaranteed by the `weak_ptr`.
> 
> Do I miss a downside of that other approach? On the plus side, it seems to be much easier to reason for the safety of that approach.

The downside is that timers are not permanently in l_Timers. Every time a timer is re-inserted into l_Timers, I need to create a valid weak_ptr. From m_Self, the -change my mind- only source at runtime. Then I had m_Self and the weak_ptr in l_Timers. I don’t see what for. All of this is safe, see my above explanation.

Reply:
> Every time a timer is re-inserted into l_Timers, I need to create a valid weak_ptr. From m_Self, the -change my mind- only source at runtime. Then I had m_Self and the weak_ptr in l_Timers. I don’t see what for.

Okay, I didn't consider that part, that would make it annoying indeed. Then keep the implementation as-is but please add a comment here why accessing members of `timer` is safe here, even if `timer->~Timer()` may already be running (which in fact, actually hasn't changed, but with that new `weak_ptr.lock()` possibly returning an empty pointer, this just looks error-prone, so a comment why it's safe doesn't hurt for sure).

Reply:
Resolved in #9738